src/detail/fastrtps_dynamic_type.cpp

```diff
@@ -142,916 +158,1190 @@ fastrtps__dynamic_type_builder_clone(
 {
-  (void) serialization_support_impl;
-  (void) allocator;
+  (void)serialization_support_impl;
+  (void)allocator;
 
-  eprosima::fastrtps::types::DynamicType_ptr type_impl_out_handle =
-    static_cast<DynamicTypeBuilder *>(type_builder_impl->handle)->build();
-  if (!type_impl_out_handle) {
+  auto type_builder_handle =
+    static_cast<const fastdds__rosidl_dynamic_typesupport_dynamic_type_builder_impl *>(
+    type_builder_impl->handle);
+  if (!type_builder_handle || !type_builder_handle->ref_type) {
+    RCUTILS_SET_ERROR_MSG("Could not get handle to type impl");
+    return RCUTILS_RET_ERROR;
+  }
+
+  auto type_handle = new (std::nothrow) fastdds__rosidl_dynamic_typesupport_dynamic_type_impl();
+  if (!type_handle) {
+    RCUTILS_SET_ERROR_MSG("Could not init new struct type");
+    return RCUTILS_RET_ERROR;
+  }
+
+  type_handle->ref_type = type_builder_handle->ref_type->build();
+  if (!type_handle->ref_type) {
     RCUTILS_SET_ERROR_MSG("Could not create dynamic type from dynamic type builder");
     return RCUTILS_RET_BAD_ALLOC;
   }
 
-  // Disgusting, but unavoidable... (we can't easily transfer ownership)
-  //
-  // We're forcing the managed pointer to persist outside of function scope by moving ownership
-  // to a new, heap-allocated DynamicType_ptr (which is a shared_ptr)
-  type_impl->handle = static_cast<void *>(new DynamicType_ptr(std::move(type_impl_out_handle)));
+  type_impl->handle = type_handle;
   return RCUTILS_RET_OK;
 }
 
-
 rcutils_ret_t
-fastrtps__dynamic_type_clone(
+fastdds__dynamic_type_clone(
   rosidl_dynamic_typesupport_serialization_support_impl_t * serialization_support_impl,
   const rosidl_dynamic_typesupport_dynamic_type_impl_t * other,
   rcutils_allocator_t * allocator,
   rosidl_dynamic_typesupport_dynamic_type_impl_t * type_impl)
 {
+  (void)serialization_support_impl;
   type_impl->allocator = *allocator;
-  auto fastrtps_impl = static_cast<fastrtps__serialization_support_impl_handle_t *>(
-    serialization_support_impl->handle);
 
-  auto type_impl_handle = eprosima::fastrtps::types::DynamicType_ptr(
-    *static_cast<const eprosima::fastrtps::types::DynamicType_ptr *>(other->handle));
-  if (!type_impl_handle) {
+  auto type_impl_handle =
+    static_cast<const fastdds__rosidl_dynamic_typesupport_dynamic_type_impl *>(other->handle);
+  if (!type_impl_handle || !type_impl_handle->ref_type) {
     RCUTILS_SET_ERROR_MSG("Could not get handle to type impl");
     return RCUTILS_RET_INVALID_ARGUMENT;
   }
 
-  eprosima::fastrtps::types::DynamicType_ptr type_impl_out_handle =
-    fastrtps_impl->type_factory_->create_alias_type(type_impl_handle, type_impl_handle->get_name());
+  auto type_impl_out_handle =
+    new (std::nothrow) fastdds__rosidl_dynamic_typesupport_dynamic_type_impl();
   if (!type_impl_out_handle) {
-    RCUTILS_SET_ERROR_MSG("Could not clone struct type");
+    RCUTILS_SET_ERROR_MSG("Could not init new struct type");
     return RCUTILS_RET_ERROR;
```

Comment:
I know that this is existing code, but this should be a BAD_ALLOC:

```suggestion
    return RCUTILS_RET_BAD_ALLOC;
```

Reply:
Done in 2ec21b3